### PR TITLE
Mesh shader

### PR DESCRIPTION
**Pull request by gpx1000**
## Description

This is the most basic of mesh shading examples.  Another PR is incoming with the LOD sample.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

**Comment by SaschaWillems** [comment]
Sadly doesn't work for me. On my RTX 2060 (Win 10) with the latest Vulkan Developer Drivers I'm only seeing a black screen and no triangle.

I do get two validation errors with the latest SDK:

```[error] [framework\core\instance.cpp:42] 461143016 - VUID-VkGraphicsPipelineCreateInfo-stage-00727: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-stage-00727 ] Object 0: handle = 0x25aec68cd60, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x1b7c7be8 | Invalid Pipeline CreateInfo[0] State: Vertex Shader required. The Vulkan spec states: If the pipeline is being created with pre-rasterization shader state the stage member of one element of pStages must be VK_SHADER_STAGE_VERTEX_BIT (https://vulkan.lunarg.com/doc/view/1.3.231.1/windows/1.3-extensions/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-stage-00727)```

```[error] [framework\core\instance.cpp:42] 987961089 - VUID-VkPipelineShaderStageCreateInfo-stage-02091: Validation Error: [ VUID-VkPipelineShaderStageCreateInfo-stage-02091 ] Object 0: handle = 0x25aec68cd60, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x3ae31701 | Invalid Pipeline CreateInfo[0] State: Mesh Shader not supported. The Vulkan spec states: If the meshShaders feature is not enabled, stage must not be VK_SHADER_STAGE_MESH_BIT_EXT (https://vulkan.lunarg.com/doc/view/1.3.231.1/windows/1.3-extensions/vkspec.html#VUID-VkPipelineShaderStageCreateInfo-stage-02091)```

Though the code mentions that these are false positives.

File: samples/extensions/mesh_shading/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+get_filename_component(FOLDER_NAME ${CMAKE_CURRENT_LIST_DIR} NAME)
+get_filename_component(PARENT_DIR ${CMAKE_CURRENT_LIST_DIR} PATH)
+get_filename_component(CATEGORY_NAME ${PARENT_DIR} NAME)
+
+add_sample_with_tags(
```
Comment (SaschaWillems) [inline samples/extensions/mesh_shading/CMakeLists.txt:22]: can you add the shaders in here? That way you can directly open them from the IDE. See e.g. https://github.com/KhronosGroup/Vulkan-Samples/blob/master/samples/extensions/raytracing_basic/CMakeLists.txt#L28

**Comment by gpx1000** [comment]
Yeah, I think we need to wait until https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/4524 is merged and released to the SDK before we accept this PR or the upcoming PR on mesh shaders.

**Comment by SaschaWillems** [comment]
Found the culprit for the triangle not being displayed:

```cpp
vkCmdDrawMeshTasksEXT(draw_cmd_buffers[i], num_workgroups_x, num_workgroups_y, 0);
```

`groupCountZ` must be at least one, or no tasks are dispatched.

Setting it to 1 makes the triangle display for me.

File: samples/extensions/mesh_shading/mesh_shading.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Basic example for VK_EXT_mesh there is only a mesh shader and a fragment shader.
```
Comment (SaschaWillems) [inline samples/extensions/mesh_shading/mesh_shading.cpp]: Should be VK_EXT_mesh_shader

File: samples/extensions/mesh_shading/mesh_shading.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Basic example for VK_EXT_mesh there is only a mesh shader and a fragment shader.
```
Comment (SaschaWillems) [inline samples/extensions/mesh_shading/mesh_shading.h]: Should be VK_EXT_mesh_shader

**Comment by gpx1000** [comment]
> Found the culprit for the triangle not being displayed:
> 
> ```c++
> vkCmdDrawMeshTasksEXT(draw_cmd_buffers[i], num_workgroups_x, num_workgroups_y, 0);
> ```
> 
> `groupCountZ` must be at least one, or no tasks are dispatched.
> 
> Setting it to 1 makes the triangle display for me.

Thanks for that!  update made and tested as working here.

File: samples/extensions/mesh_shading/mesh_shading.cpp [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+ */
+
+#include "mesh_shading.h"
+
+MeshShading::MeshShading() :
+    pipeline(VK_NULL_HANDLE), pipeline_layout(VK_NULL_HANDLE), descriptor_set(VK_NULL_HANDLE), descriptor_set_layout(VK_NULL_HANDLE)
+{
+	title = "Mesh shading";
+
+	// vk_mesh_ext requires Vulkan 1.3 and device properties 2.  SPIR-V must also be set to at least 1.4.
```
Comment (SaschaWillems) [inline samples/extensions/mesh_shading/mesh_shading.cpp]: Oddly the spec is not clear on this (see https://github.com/KhronosGroup/Vulkan-Docs/issues/1976). VK_EXT_Mesh_Shader only requires VK 1.0, but SPIRV 1.4 requires 1.1. So the baseline is currently 1.1. Maybe we can lower this for our sample?

File: samples/extensions/mesh_shading/mesh_shading.cpp [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+	// Submit to queue
+	VK_CHECK(vkQueueSubmit(queue, 1, &submit_info, VK_NULL_HANDLE));
+
+	ApiVulkanSample::submit_frame();
+}
+
+void MeshShading::prepare_pipelines()
+{
+	std::vector<VkDescriptorPoolSize> pool_sizes;
```
Comment (asuessenbach) [inline samples/extensions/mesh_shading/mesh_shading.cpp]: Are the `pool_sizes` needed at all here?

File: samples/extensions/mesh_shading/mesh_shading.cpp [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+	VkDescriptorPoolCreateInfo        descriptor_pool_create_info =
+	    vkb::initializers::descriptor_pool_create_info(
+	        static_cast<uint32_t>(pool_sizes.size()),
+	        pool_sizes.data(),
+	        2);
+
+	VK_CHECK(vkCreateDescriptorPool(get_device().get_handle(), &descriptor_pool_create_info, nullptr, &descriptor_pool));
+
+	std::vector<VkDescriptorSetLayoutBinding> set_layout_bindings;
```
Comment (asuessenbach) [inline samples/extensions/mesh_shading/mesh_shading.cpp]: Are the `set_layout_bindings` needed here?

File: shaders/mesh_shading/ms.mesh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;
+layout(triangles, max_vertices = 64, max_primitives = 124) out;
```
Comment (asuessenbach) [inline shaders/mesh_shading/ms.mesh]: Even though it probably doesn't matter in this sample, you should keep max_vertices and max_primitives at a reasonable level (here: 3 and 1).

File: samples/extensions/mesh_shading/mesh_shading.cpp (hunk 4 above)
Comment (SaschaWillems) [reply samples/extensions/mesh_shading/mesh_shading.cpp]: The spec issue has been resolved. VK_EXT_mesh_shader requires Vulkan 1.1, so we should be able to lower the requirement here.

Comment (gpx1000) [reply samples/extensions/mesh_shading/mesh_shading.cpp]: Thanks for chasing that down I lowered it to 1.1

**Review by asuessenbach — DISMISSED** [review]

**Review by SaschaWillems — DISMISSED** [review]
Works fine for me on Windows 10 with an RTX 2060. 

Only thing I'm wondering: Do we want a more detailed readme?

**Comment by gpx1000** [comment]
I was debating that; but given the simplicity of the sample I'm not sure how much a more detailed README would help.  I might spend some time today thinking it over though.

**Pull request merged**